### PR TITLE
use flexbox

### PR DESCRIPTION
**Pull request by sethladd**

### CODE REVIEW

**Comment by devoncarew** [comment]
@sethladd, it sounds like this is ready for review now? If so, I'll look through some of the css changes; /cc @jcollins-g for more general review.

(preview available here: https://sethladd.github.io/test-dartdoc-flexbox/)

**Comment by devoncarew** [comment]
For the travis failure, I think you need to regenerate the example docs - `grind update-test-package-docs`.

File: tool/grind.dart [hunk 1]
```diff
@@ -188,7 +188,7 @@ test() {
 
 @Task('Generate docs for dartdoc')
 testDartdoc() {
-  delete(docsDir);
+  //delete(docsDir);
```
Comment (devoncarew) [inline tool/grind.dart]: Intentional?

File: lib/resources/styles.css [hunk 2]
```diff
@@ -10,15 +10,84 @@
...
+  justify-content: flex-end;
+  padding-right: 30px;
+  
+}
+footer {
+  flex: 0 0 16px;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  font-size: 8px;
```
Comment (devoncarew) [inline lib/resources/styles.css]: I think we want to strike this - it makes the footer text very small.

We probably also want `padding: 20px;` to better match dartlang.org.

File: tool/grind.dart (hunk 1 above)
Comment (jcollins-g) [inline tool/grind.dart]: this needs to be uncommented, I think

File: lib/resources/styles.css [hunk 3]
```diff
@@ -10,15 +10,84 @@
...
 }
 
+.main-content p {
+  max-width: 700px;
+}
+
+.sidebar-offcanvas-right {
+  flex: 0 1 12em;
+  overflow-y: auto;
+  padding: 20px 15px 15px 15px;
```
Comment (devoncarew) [inline lib/resources/styles.css]: I think also `margin-right: 20px;` for space on the right side of the page.

File: tool/grind.dart (hunk 1 above)
Comment (sethladd) [inline tool/grind.dart]: yeah, for testing. will remove for final PR :)

**Comment by sethladd** [comment]
> And as Devon noted, regenerate test docs.

Roger that!

File: lib/resources/styles.css (hunk 3 above)
Comment (sethladd) [inline lib/resources/styles.css]: can do

File: lib/resources/styles.css (hunk 2 above)
Comment (sethladd) [inline lib/resources/styles.css]: I'm not sure about that. our footer is always visible. Adding 20px padding is a lot of wasted space.

**Comment by sethladd** [comment]
Thanks! I'll get the PR into tip-top shape.

**Comment by devoncarew** [comment]
For some of the elements (`.sidebar`, `.main-content`) we may want to prefer top and bottom margin in preference to top and bottom padding - this way there's some space between the scroll bar and the header and footer areas (visually, it will look like we're separating with spacing rather than lines).

```
.sidebar,
.main-content {
  margin: 20px 0;
}
```

If we do this, we'd need to adjust the padding for these elements, and for the contained `ol` elements).

**Comment by devoncarew** [comment]
The new layout looks great :)

I made a pass through, visually diffing against dartlang.org (not necessarily to match dartlang exactly, but more for aesthetics).

File: lib/resources/styles.css (hunk 2 above)
Comment (devoncarew) [inline lib/resources/styles.css]: hmm, the different - much smaller font - is distracting, and there's not enough separation between the content and its border.

**Comment by sethladd** [comment]
> hmm, the different - much smaller font - is distracting, and there's not enough separation between the content and its border.

Can you play with the footer and provide a specific set of CSS that you or I will merge. Sounds like you have a stronger opinion here, so to avoid a lot of back and forth, send me the CSS for the footer. :) Thanks!

**Comment by devoncarew** [comment]
This should do it:

```
footer {
  flex: 0 0 16px;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 20px;
}
```

> I'm not sure about that. our footer is always visible.

Having is scroll off the screen would also work, but likely more complicated w/ the layout changes.

**Comment by sethladd** [comment]
> Having is scroll off the screen would also work

Yeah, that's what I want, but can't do it unless larger changes. I think it requires JavaScript (because of the three independently scrolling columns now)

**Comment by jcollins-g** [comment]
I don't feel strongly regardless, but a compromise between what you and Devon are trying to achieve might include putting the footer only in the middle column.  Not exactly what we want (what I want is the magically appearing footer when you scroll down referenced in #1350), but it's reasonably close.

File: lib/resources/styles.css (hunk 3 above)
Comment (sethladd) [inline lib/resources/styles.css]: done

**Comment by sethladd** [comment]
@devoncarew @jcollins-g I think it's ready. https://sethladd.github.io/test-dartdoc-flexbox/ was updated. Tests are passing. Footer was embiggened.

PTAL, thanks!

**Comment by devoncarew** [comment]
re-reviewing https://sethladd.github.io/test-dartdoc-flexbox/, the persistent black footer is a bit much. I'll play with some options after this PR lands -

**Comment by sethladd** [comment]
> the persistent black footer is a bit much

That's why I was trying to make it smaller :)

Thanks for all the reviews and help!

**Pull request merged**